platform/linux-generic/pktio/socket_xdp.c

```diff
@@ -0,0 +1,591 @@
+		return 0;
+
+	priv = pkt_priv(pktio_entry);
+	pool = priv->sock_info.umem_info->pool;
+	odp_ticketlock_lock(&priv->tx_lock);
+
+	for (i = 0; i < num; ++i) {
+		odp_pkt = ODP_PACKET_INVALID;
+
+		if (odp_unlikely(odp_packet_num_segs(packets[i])) > 1) {
+			/* TODO: handle segmented packets */
+			ODP_ERR("Only single-segment packets supported\n");
+			break;
+		}
+
+		odp_hdr = packet_hdr(packets[i]);
+
+		if (odp_hdr->event_hdr.pool_ptr != pool_entry_from_hdl(pktio_entry->s.pool)) {
+			odp_pkt = odp_packet_copy(packets[i], pool->pool_hdl);
+
+			if (odp_unlikely(odp_pkt == ODP_PACKET_INVALID)) {
+				ODP_ERR("Unable to allocate packet for sending\n");
+				break;
+			}
+
+			odp_hdr = packet_hdr(odp_pkt);
+		}
+
+		if (xsk_ring_prod__reserve(&priv->sock_info.tx, 1U, &start_idx) == 0U) {
+			handle_pending_tx(&priv->sock_info, NUM_XDP_DESCS);
+
+			if (xsk_ring_prod__reserve(&priv->sock_info.tx, 1U, &start_idx) == 0U) {
+				if (odp_pkt != ODP_PACKET_INVALID)
+					odp_packet_free(odp_pkt);
+
+				break;
+			}
+		}
+
+		if (odp_pkt != ODP_PACKET_INVALID)
+			odp_packet_free(packets[i]);
+
+		populate_tx_desc(pool, odp_hdr, xsk_ring_prod__tx_desc(&priv->sock_info.tx,
+								       start_idx));
+	}
+
+	xsk_ring_prod__submit(&priv->sock_info.tx, i);
+	handle_pending_tx(&priv->sock_info, NUM_XDP_DESCS);
+	odp_ticketlock_unlock(&priv->tx_lock);
+
+	return i;
+}
+
+static uint32_t sock_xdp_mtu_get(pktio_entry_t *pktio_entry)
+{
+	return pkt_priv(pktio_entry)->sock_info.mtu;
+}
+
+static int sock_xdp_mtu_set(pktio_entry_t *pktio_entry, uint32_t maxlen_input,
+			    uint32_t maxlen_output ODP_UNUSED)
+{
+	pkt_xdp_t *priv = pkt_priv(pktio_entry);
+	int ret;
+
+	ret = _odp_mtu_set_fd(priv->sock_info.helper_sock, pktio_entry->s.name, maxlen_input);
+	if (ret)
+		return ret;
+
+	priv->sock_info.mtu = maxlen_input;
+
+	return 0;
+}
+
+static int sock_xdp_promisc_mode_set(pktio_entry_t *pktio_entry,  int enable)
+{
+	return _odp_promisc_mode_set_fd(pkt_priv(pktio_entry)->sock_info.helper_sock,
+					pktio_entry->s.name, enable);
+}
+
+static int sock_xdp_promisc_mode_get(pktio_entry_t *pktio_entry)
+{
+	return _odp_promisc_mode_get_fd(pkt_priv(pktio_entry)->sock_info.helper_sock,
+					pktio_entry->s.name);
+}
+
+static int sock_xdp_mac_addr_get(pktio_entry_t *pktio_entry ODP_UNUSED, void *mac_addr)
+{
+	return _odp_mac_addr_get_fd(pkt_priv(pktio_entry)->sock_info.helper_sock,
+				    pktio_entry->s.name, mac_addr) ? -1 : ETH_ALEN;
+}
+
+static int sock_xdp_link_status(pktio_entry_t *pktio_entry)
+{
+	return _odp_link_status_fd(pkt_priv(pktio_entry)->sock_info.helper_sock,
+				   pktio_entry->s.name);
+}
+
+static int sock_xdp_link_info(pktio_entry_t *pktio_entry, odp_pktio_link_info_t *info)
+{
+	return _odp_link_info_fd(pkt_priv(pktio_entry)->sock_info.helper_sock,
+				 pktio_entry->s.name, info);
+}
+
+static int sock_xdp_capability(pktio_entry_t *pktio_entry, odp_pktio_capability_t *capa)
+{
+	pkt_xdp_t *priv = pkt_priv(pktio_entry);
```

Comment:
Could be zeroed to be on the safe side. Looks like `min_output_queue_size` and `max_output_queue_size` are currently not initialized.

Reply:
v9.